packages/indexer-common/src/allocations/query-fees.ts

```diff
@@ -367,14 +410,177 @@ export class AllocationReceiptCollector implements ReceiptCollector {
+    pendingRAVs: Eventual<RavWithAllocation[]>,
+  ): Eventual<ValidRavs> {
+    return pendingRAVs.tryMap(
+      async (pendingRAVs) => {
+        return await pReduce(
+          pendingRAVs,
+          async (results, rav) => {
+            // allocationId doesn't have 0x prefix, but still matched in contract calls
+            // destory voucher if AllocationIDTracker.isAllocationIdUsed(rav.allocationId)
+            if (
+              BigNumber.from(rav.rav.rav.valueAggregate).lt(
+                this.voucherRedemptionThreshold,
+              )
+            ) {
+              results.belowThreshold.push(rav)
+            } else {
+              results.eligible.push(rav)
+            }
+            return results
+          },
+          { belowThreshold: <RavWithAllocation[]>[], eligible: <RavWithAllocation[]>[] },
+        )
+      },
+      { onError: (err) => this.logger.error(`Failed to reduce to signed RAVs`, { err }) },
+    )
+  }
+
+  // redeem only if last is true
+  // Later can add order and limit
+  protected async pendingRAVs(): Promise<ReceiptAggregateVoucher[]> {
+    let ravsToRedeem: ReceiptAggregateVoucher[] = []
+    const unfinalizedRAVs = await this.models.receiptAggregateVouchers.findAll({
+      where: { last: true, final: false },
+    })
+    // Obtain allocationIds to use as filter in subgraph
+    const allocationIdsAddress = unfinalizedRAVs.map(
+      (rav) => rav.getSignedRAV().rav.allocationId.toLowerCase(),
+    )
+    const allocationIds = unfinalizedRAVs.map(
+      (rav) => rav.getSignedRAV().rav.allocationId.toLowerCase().replace('0x', ''),
+    )
+
+    if (allocationIdsAddress.length > 0) {
+      console.log(`Obtaining redeemed RAVs against subgraph`)
+      const redeemedRavsQueryResp = await this.tapSubgraph.query(
+        gql`
+          query transactions($allocationIds: [String!]!) {
+            transactions(
+              where: { type: "redeem", allocationID_in: $allocationIds }
+            ) {
+              allocationID
+            }
+          }
+        `,
+        { allocationIdsAddress },
+      )
+      // From obtained allocationIDs, get those that were not returned and redeem
+      const returnedAllocationIDs = redeemedRavsQueryResp.data.transactions.map(
+        (transaction) => transaction.allocationID,
+      )
+      const nonRedeemedAllocationIDs = allocationIdsAddress.filter(
+        (allocationID) => !returnedAllocationIDs.includes(allocationID),
+      )
+      ravsToRedeem = unfinalizedRAVs.filter((rav) =>
+        nonRedeemedAllocationIDs.includes(rav.getSignedRAV().rav.allocationId),
+      )
+
+      // Update those that redeemed_at is older than 20 minutes and mark as final
+      const finalizeTimeFrame = Date.now() - 20 * 60 * 1000
+      const Op = Sequelize.Op
+      await this.models.receiptAggregateVouchers.update(
+        { final: true },
+        {
+          where: {
+            allocationId: { [Op.in]: allocationIds },
```

Comment:
You can use  `where: { last: true, final: false,  redeemedAt: { [Op.lt]: new Date(finalizeTimeFrame) }}`

Reply:
it already does that

Reply:
I think you don't need to filter by allocationId. Any RAV that remained redeemed on chain for more than N minutes can be marked as final. Even safer if you do https://github.com/graphprotocol/indexer/pull/869#discussion_r1526944344 I think.

Reply:
It does that, but you are sending all `allocationIds` which is inefficient. You can simplify that by using just last, final and redeemedAt